docs: add steps for in-place upgrades (#6963)
confluentinc/ksql: docs/operate-and-deploy/installation/upgrading.md

cluster on the new version, porting across your database schema, and finally thinking about your data.
Read on for details.

> ## (1) In-Place upgrade

Reviewer: ```suggestion
### In-Place upgrade
```
By changing the subheading level, this section is now a subsection of the "How to upgrade" section above, rather than at the same level.

> #### Steps :-

Reviewer: ```suggestion
#### Steps
```
No colons necessary after headings in markdown. Same for below.

1. Download the [new ksqlDB version](https://docs.confluent.io/platform/current/installation/installing_cp/zip-tar.html#get-the-software) to upgrade to.

> 2. Set it up with all required configs. In our case we mostly copied all configs from the current version(which is running) to the new version.

Reviewer: ```suggestion
2. Set it up with all required configs. For the vast majority of upgrades, you may simply copy existing configs over to the new version.
```

> 3. Take Backup and then copy all UDF JAR to new version .(OPTIONAL STEP In case you don't have any UDF).

Reviewer: ```suggestion
3. Copy source UDFs and recompile against the new ksqlDB version, to accommodate any changes to the UDF framework.
```
Recompiling UDFs is not necessary for most upgrades, but sometimes it is so it doesn't hurt to include a mention in these upgrade steps.

> 4. Stop ksqlDB process of current version(5.5 in our case) and start newer version(6 in our case) with latest KsqlDB config file which you created in step2.

Reviewer: ```suggestion
4. Stop ksqlDB process of current version and start newer version with latest ksqlDB config file which you created in step 2.
```
Minor rewrite to remove parentheticals.

Author: @vcrfxia actually i missed all these comments, i have made all suggested minor changes.

Author: Let me know if something is missing.

5. KsqlDB will start executing queries and rebuild all states using command topic because the value of "ksql.service.id" is the same.
6. Now KsqlDB Server with the latest version is up and running.

#### Troubleshooting an Upgrade :-

1. If the ksqlDB server fails to start with an error message such as "KsqlException: UdfFactory not compatible with existing factory. function: <FUNCTION_NAME> existing". 

> This error is because Confluent KsqlDB 6 has these functions as in built so it will create conflict with your UDFs. So to solve this you can create a new udf jar by eliminating those inbuilt functions and deploy it and then try to start KsqlDB again.

Reviewer: ```suggestion
This error is because the new ksqlDB version has introduced an in-built function of the same name, creating a conflict with your UDF. To solve this you can create a new UDF jar by renaming or eliminating the conflicting function(s), and restarting ksqlDB with the new UDF jar.
```
Minor rewrite to remove mention of a specific ksqlDB version, and clarify that the conflict is a name conflict :)

2. "desribe extended" command not working :- This generally happens when ksqlCli and ksqlServer version is not same. Generally people create alias for ksqlCli so make sure to change it to newer version. So all an all, version of ksqlCli and ksqlServer should be same.

> ## (2) Steps to follow when "In-Place" upgrade is not supported

Reviewer: ```suggestion
### Steps to follow when "In-Place" upgrade is not supported
```
See above regarding subheading level. This means we'll also have to update the subheadings below ("Port the database schema", "Rebuild state", and "Destroy the old cluster") to be one level lower.

cluster on the new version, porting across your database schema, and finally thinking about your data.
Read on for details.

### In-Place upgrade

> #### Steps

Reviewer 2: ```suggestion
Follow these steps to do an in-place upgrade of ksqlDB.
```

1. Download the [new ksqlDB version](https://docs.confluent.io/platform/current/installation/installing_cp/zip-tar.html#get-the-software) to upgrade to.
2. Set it up with all required configs. For the vast majority of upgrades, you may simply copy existing configs over to the new version.
3. Copy source UDFs and recompile against the new ksqlDB version, to accommodate any changes to the UDF framework.

> 4. Stop ksqlDB process of current version and start newer version with latest ksqlDB config file which you created in step 2.

Reviewer 2: ```suggestion
4. Stop the ksqlDB process running the current version and start the newer version with the latest ksqlDB config file, which you created in Step 2.
```

> 5. KsqlDB will start executing queries and rebuild all states using command topic because the value of "ksql.service.id" is the same.

Reviewer 2: ```suggestion
5. ksqlDB starts executing queries and rebuilds all states, using the command topic, because the value of `ksql.service.id` is the same as the previous version.
```

> 6. Now KsqlDB Server with the latest version is up and running.

Reviewer 2: ```suggestion

```

> #### Troubleshooting an Upgrade :-

Reviewer 2: ```suggestion
#### Troubleshooting an upgrade
```

Author: Addressed suggested changes.

> 1. If the ksqlDB server fails to start with an error message such as "KsqlException: UdfFactory not compatible with existing factory. function: <FUNCTION_NAME> existing". 

Reviewer 2: Maybe a bullet list instead of numbered?

> 1. If the ksqlDB server fails to start with an error message such as "KsqlException: UdfFactory not compatible with existing factory. function: <FUNCTION_NAME> existing". 

Reviewer 2: ```suggestion
ksqlDB Server fails to start with an error message like "KsqlException: UdfFactory not compatible with existing factory. function: <FUNCTION_NAME> existing": 
```

> This error is because the new ksqlDB version has introduced an in-built function of the same name, creating a conflict with your UDF. To solve this you can create a new UDF jar by renaming or eliminating the conflicting function(s), and restarting ksqlDB with the new UDF jar.

Reviewer 2: ```suggestion
This error occurs when the new ksqlDB version introduces a built-in function that has the same name as yours, creating a conflict with your UDF. Create a new UDF jar by renaming or eliminating the conflicting function(s), and restart ksqlDB with the new UDF jar.
```

> 2. "desribe extended" command not working :- This generally happens when ksqlCli and ksqlServer version is not same. Generally people create alias for ksqlCli so make sure to change it to newer version. So all an all, version of ksqlCli and ksqlServer should be same.

Reviewer 2: ```suggestion
DESCRIBE EXTENDED command doesn't work: 
Usually, this happens when the ksqlDB CLI and ksqlDB Server versions are different. You can create an alias for ksqlDB CLI to make sure to change it to newer version. You should ensure that the ksqlDB CLI and Server versions are the same.
```

Reviewer 2: Not sure I understand the "alias" guidance. Maybe provide an example command?

cluster on the new version, porting across your database schema, and finally thinking about your data.
Read on for details.

### In-Place upgrade

Follow these steps to do an in-place upgrade of ksqlDB.

1. Download the [new ksqlDB version](https://docs.confluent.io/platform/current/installation/installing_cp/zip-tar.html#get-the-software) to upgrade to.
2. Set it up with all required configs. For the vast majority of upgrades, you may simply copy existing configs over to the new version.
3. Copy source UDFs and recompile against the new ksqlDB version, to accommodate any changes to the UDF framework.
4. Stop the ksqlDB process running the current version and start the newer version with the latest ksqlDB config file, which you created in Step 2.
5. ksqlDB starts executing queries and rebuilds all states, using the command topic, because the value of `ksql.service.id` is the same as the previous version. ksqlDB Server with the latest version is up and running.

> ### Troubleshooting an Upgrade :-

Reviewer: ```suggestion
### Troubleshooting an Upgrade
```

Author: Done.

### `ALL`, `WINDOWSTART` and `WINDOWEND` are now reserved identifiers

Any query using these identifiers will need to be changed to either use some other identifier, or
to quote them.

> ### Upgrading from confluentksqlDB 5.5 to confluentksqlDB 6

Reviewer: These details for in-place upgrades are applicable for all in-place upgrades, not just from 5.5 to 6.0. How about rewriting them to be generic from `<current_version>` to `<new_version>` and moving this section to be part of the section titled "How to upgrade" above? We can have two sub-sections under "How to upgrade": "In-place upgrades" (which you've provided the content for) and "Upgrades that cannot be performed in place" (the existing content under that section).

Author: Understood and changes done in my branch.

In-place upgrades are supported from confluentksqlDB 5.5 to confluentksqlDB 6.

> #### Steps of "In-Place" upgrade :-

Reviewer 2: ```suggestion
#### In-place upgrade
```

> 1. Download the latest confluent version.

Reviewer: ```suggestion
1. Download the new ksqlDB version to upgrade to.
```
The new version need not be a Confluent ksqlDB version, the community versions work as well with the same instructions.

Reviewer 2: Might be helpful to provide a link.

Author: @JimGalasyn I have added a download link. Can you check if that makes sense ?

2. Set it up with all required configs. In our case we mostly copied all configs from the current version(which is running) to the new version.
3. Take Backup and then copy all UDF JAR to new version .(OPTIONAL STEP In case you don't have any UDF).
4. Stop ksqlDB process of current version(5.5 in our case) and start newer version(6 in our case) with latest KsqlDB config file which you created in step2.
5. KsqlDB will start executing queries and rebuild all states using command topic because the value of "ksql.service.id" is the same.
6. Now KsqlDB Server with the latest version is up and running.

> #### Common Issues :-

Reviewer: How about we rename this section "Troubleshooting an Upgrade"?

Author: Understood and changes done in my branch.

> 1. There might have been some UDFs like ARRAY_EXCEPT , ARRAY_LENGTH , ARRAY_REMOVE etc since they were not available with Confluent KsqlDB 5.5 so when you start newer version, it will give an error saying "KsqlException: UdfFactory not compatible with existing factory. function: ARRAY_EXCEPT existing". 

Reviewer: To rewrite this bullet to be generic for upgrades between different versions, we can say something like "If the server fails to start with an error message such as `KsqlException: UdfFactory not compatible with existing factory. function: <FUNCTION_NAME> existing`, that means the name of one of your UDFs conflicts with a newly added built-in function of the same name in the new version. To solve this..."

Author: This makes sense.
Changes done.

This error is because Confluent KsqlDB 6 has these functions as in built so it will create conflict with your UDFs. So to solve this you can create a new udf jar by eliminating those inbuilt functions and deploy it and then try to start KsqlDB again.

> 2. "desribe extended" command not working :- This generally happens when ksqlCli and ksqlServer version is not same. Generally people create alias for ksqlCli so make sure to change it to newer version. So all an all, version of ksqlCli and ksqlServer should be same.

Reviewer: You may be glad to hear that we are working to improve this experience going forward :) https://github.com/confluentinc/ksql/issues/3860
